### PR TITLE
Implement a virtual value stack for JIT-compiled functions

### PR DESCRIPTION
**Pull request by aviansie-ben**
This PR adds support for a custom virtual stack that should eliminate most unnecessary reads and writes to/from the value stack that were caused by the previous naive implementation (#24). This should hopefully give a considerable performance boost.

### CODE REVIEW

**Comment by aviansie-ben** [comment]
A lot more work has been put into making the virtual value stack work. The new implementation uses `VirtualMachineState` rather than rolling our own logic for handling merge points and state propagation between `BytecodeBuilder`s. The `VirtualMachineOperandStack` provided by JitBuilder is not being used, as it makes a number of assumptions about the virtual stack that we don't follow (e.g. it wants to be able to commit and reload the value stack to/from memory and thus requires that all elements of the value stack to be of the same type).

@Leonardo2718 This should be ready for more thorough review and all opcodes are now implemented. I will however be running some additional stress and performance tests on this before this should be merged.

File: src/jit/function-builder.h
Comment (Leonardo2718) [inline src/jit/function-builder.h]: Would it be safer to use `dynamic_cast` here?

File: src/jit/function-builder.h
Comment (Leonardo2718) [inline src/jit/function-builder.h]: Relatively minor nitpick: In every other `Emit*` helper, the `stack` argument comes after the `pc` argument. Any reason for this one to be different?

File: src/jit/stack.h
Comment (Leonardo2718) [inline src/jit/stack.h]: I'd use `override` instead of `virtual` here.

File: src/jit/function-builder.h
Comment (Leonardo2718) [inline src/jit/function-builder.h]: I'd use `override` instead of `virtual` here.

File: src/jit/stack.h
Comment (Leonardo2718) [inline src/jit/stack.h]: Would it be safer to use `dynamic_cast` here?

File: src/jit/stack.h
Comment (Leonardo2718) [inline src/jit/stack.h]: I'd use `override` of `virtual` here.

File: src/jit/stack.cc
Comment (Leonardo2718) [inline src/jit/stack.cc:42]: Can this condition be simplified as `keep_count >= 0 && values_size() > drop_count + keep_count`?

File: src/jit/stack.cc
Comment (Leonardo2718) [inline src/jit/stack.cc]: Nice! To avoid potential symbol conflicts later on, can this function be made `static`?

File: src/jit/stack.cc
Comment (Leonardo2718) [inline src/jit/stack.cc:69]: I wonder if we can use a standard algorithm here instead of a hand-rolled loop, for example (untested):

```c++
std::transform(values_.begin(), values_end(),
               other->values_.begin(),
               other->values_.begin(),
[&](TR::IlValue* v, TR::IlValue* o) {
  TR_ASSERT(v && o, "Attempt to merge divergent stacks");
  TR_ASSERT((*this_it)->getDataType() == (*other_it)->getDataType(),
            "Attempt to merge divergent stacks");
  b->StoreOver(o,v);
  return o;
}
);
```

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc:991]: (This is outside the scope of this PR, but we really should consider caching all these `typeDictionary()->PointerTo(*)` as locals to simplify the code.)

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc]: This change introduces a subtlety that, I think, is worth giving some more consideration.

Obviously, the `jit_meta` information is needed to setup the stack frame for the call. However, the above condition will prevent compilation of functions that _call_ any other function for which there is no `jit_meta` entry. In practice, I don't believe it will be a (performance) problem because the table should have entries for all functions in a module.

Nevertheless, it's a limitation worth documenting as it may affect future changes.

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc]: Conceptually, I think I understand the point of `v`. But when `StoreOver(v, true_value)` is called later, I think it "hides" the fact that `v` actually points to the same object as `false_value`.

I think just using `false_value` would make the subtlety more obvious and better communicate what the code is actually doing.

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc]: Can the name of the type be printed as part of the assert message here? ex:

```c++
TR_ASSERT_FATAL(false, "Invalid primitive type %s", dt.toString());
```

File: src/jit/stack.cc
Comment (Leonardo2718) [inline src/jit/stack.cc]: Can the values of `*this_it` and `*other_it` be printed as part of the assert message?

File: src/jit/stack.cc
Comment (Leonardo2718) [inline src/jit/stack.cc]: Can the names of the data types be printed as part of the assert message?

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc]: I'm not a fan of the names `PushPhys` and `PopPhys` as they don't fully convey what these functions do. Maybe `MoveToPhysStack` and `MoveFromPhysStack` would work better (I'm open to suggestions :slightly_smiling_face:)?

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc]: Can the expected and actual alloca size be printed as part of the assert message?

File: src/jit/function-builder.cc
Comment (Leonardo2718) [inline src/jit/function-builder.cc]: Would it not be better to just pass `state->stack`?

**Review by Leonardo2718 — CHANGES_REQUESTED** [review]
Looks like an excellent start :slightly_smiling_face: .

File: src/jit/function-builder.cc
Comment (aviansie-ben) [reply src/jit/function-builder.cc]: At this point in the function `state` is stale (it refers to the state right before the first `BytecodeBuilder`), so directly replacing this with `state->stack` would be incorrect. I agree that this isn't very pretty, though. I think I'll go ahead and add a reload of `state` from `work_item.builder->vmState()` and then use `state->stack` to reduce the amount of stuff going on in this line, as this could quickly get out-of-hand if we start adding more state.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [reply src/jit/function-builder.cc]: Looking back on it, I agree that the names are no longer representative. They used to actually do what the names suggested, but they were later changed for efficiency reasons and the names were never changed. I'm fine with `MoveToPhysStack` and `MoveFromPhysStack` here.

File: src/jit/function-builder.cc
Comment (aviansie-ben) [reply src/jit/function-builder.cc]: Agreed

File: src/jit/function-builder.cc
Comment (aviansie-ben) [reply src/jit/function-builder.cc]: I'm starting to think that we might just want to use asserts instead for missing `jit_meta` entries. It honestly shouldn't happen and it'd be nice to know why performance takes a nosedive if this assumption is ever violated. Thoughts?

File: src/jit/function-builder.h
Comment (aviansie-ben) [reply src/jit/function-builder.h]: Ah, right. I sort of got used to not using it since OMR and OpenJ9 don't fully support C++11.

File: src/jit/function-builder.h
Comment (aviansie-ben) [reply src/jit/function-builder.h]: Just switching from `static_cast<WabtState*>` to `dynamic_cast<WabtState*>` wouldn't actually gain any safety, since `dynamic_cast` of a pointer type returns `nullptr` if the cast is invalid. Since dereferencing `nullptr` is UB, we wouldn't actually gain anything.

I could replace this with `&dynamic_cast<WabtState&>(*other).stack` though... Thoughts?

File: src/jit/function-builder.h
Comment (aviansie-ben) [reply src/jit/function-builder.h]: Not really. Just one of those weird inconsistencies that snuck in... I'll go ahead and fix that.

File: src/jit/stack.cc
Comment (aviansie-ben) [reply src/jit/stack.cc]: Right... I intended this to be `static` in the first place, but I guess I forgot to add it. :sweat_smile:

File: src/jit/stack.cc
Comment (aviansie-ben) [reply src/jit/stack.cc:42]: No. `keep_count >= 0` is always true since `size_t` is an unsigned type, whereas the original `drop_count + keep_count >= drop_count` is checking for unsigned overflow.

File: src/jit/stack.cc
Comment (aviansie-ben) [reply src/jit/stack.cc:69]: Hmm... Should be possible. I'll play around with it and see if I can get it to work.

File: src/jit/stack.cc
Comment (Leonardo2718) [reply src/jit/stack.cc:42]: I see.

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: Agreed.

File: src/jit/function-builder.h
Comment (Leonardo2718) [reply src/jit/function-builder.h]: I like `&dynamic_cast<WabtState&>(*other).stack`. I really just want to have something "defined" happen if the cast becomes invalid because of some change we do later on to the class hierarchy :slightly_smiling_face:.

File: src/jit/function-builder.cc
Comment (Leonardo2718) [reply src/jit/function-builder.cc]: Yeah ok, that makes sense. I somehow missed that `state` is stale :stuck_out_tongue:.

File: src/jit/stack.cc
Comment (aviansie-ben) [reply src/jit/stack.cc:69]: I've played around with this. As part of solving another problem you mentioned, I ended up adding stack offsets to the `TR_ASSERT` messages, however it seems like this change would make that more difficult since that would require keeping a separate counter for the stack offset instead of being able to use `this_it - values_.begin()`. In the interest of being able to do that easily, I'm leaning towards leaving this as-is.

Comment (Leonardo2718) [reply src/jit/stack.cc:69]: Makes sense. I agree the stack offset is useful to have around for debugging purposes.

**Comment by aviansie-ben** [comment]
Most feedback has been addressed. Currently, I'm running into codegen problems in the raytracer. Looks like it might be an upstream issue with OMR...

**Comment by Leonardo2718** [comment]
Alright, I won't merge until WIP is removed from the PR title :slightly_smiling_face: .

Once you have a reduced test case for reproducing the codegen problem, can you open an issue against OMR? If you can't create a minimal test case, it's probably ok to just open an issue with whatever you have.

**Comment by aviansie-ben** [comment]
At this point, I've found a minimal test case for the codegen issue I was running into, but due to our current situation I can't test against upstream OMR to see if the problem still exists. In order to do that, we'll need to change our usage of the custom `CoerceTo` service to use the new `ConvertBitsTo` service and update to use the new JitBuilder client API. Due to the intrusive nature of these changes, I don't think performing those changes while this PR is unmerged would be a good idea.

In light of this, my current plan is to have these changes merged then begin work on updating to use the latest version of upstream OMR. Then, if the problem still exists we can open an upstream issue about it. Since this is the only known issue with this changeset right now, I'll be removing the WIP marker so that it can be merged.

@Leonardo2718 Please complete your review and, if appropriate, merge this PR at your earliest convenience.

**Review by Leonardo2718 — APPROVED** [review]

**Pull request merged**
